clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -1061,16 +1082,36 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
                 requestRejoin();
                 future.raise(error);
             } else if (error == Errors.ILLEGAL_GENERATION) {
-                log.info("Attempt to heartbeat failed since generation {} is not current", sentGeneration.generationId);
-                resetGenerationOnResponseError(ApiKeys.HEARTBEAT, error);
-                future.raise(error);
+                if (generationUnchanged()) {
+                    log.info("Attempt to heartbeat failed since current {} is not valid, resetting generation", sentGeneration);
+                    resetGenerationOnResponseError(ApiKeys.HEARTBEAT, error);
+                    future.raise(error);
+                } else {
+                    // if the generation has changed, then ignore this error
+                    log.info("Attempt to heartbeat failed since old {} is not valid, ignoring the error", sentGeneration);
+                    future.complete(null);
+                }
             } else if (error == Errors.FENCED_INSTANCE_ID) {
-                log.error("Received fatal exception: group.instance.id gets fenced");
-                future.raise(error);
+                if (generationUnchanged()) {
+                    log.info("Attempt to heartbeat failed since current {} gets fenced with group instance id {}",
+                        sentGeneration, rebalanceConfig.groupInstanceId);
+                    future.raise(error);
+                } else {
+                    // if the generation has changed, then ignore this error
+                    log.info("Attempt to heartbeat failed since old {} gets fenced with group instance id {}, " +
+                        "ignoring the error", sentGeneration, rebalanceConfig.groupInstanceId);
+                    future.complete(null);
+                }
             } else if (error == Errors.UNKNOWN_MEMBER_ID) {
-                log.info("Attempt to heartbeat failed since member id {} is not valid.", sentGeneration.memberId);
-                resetGenerationOnResponseError(ApiKeys.HEARTBEAT, error);
-                future.raise(error);
+                if (generationUnchanged()) {
```

Comment:
In other response handlers, we combine UNKNOWN_MEMBER_ID and ILLEGAL_GENERATION cases. Do you think we could also consolidate here as well?

Reply:
Yes, I agree we should try to consolidate. Looking at this handler alone, we have basically the same code for `ILLEGAL_GENERATION`, `UNKNOWN_MEMBER_ID`, and `FENCED_INSTANCE_ID`.

Reply:
SG, let me try it out.